plugin/pkg/admission/webhook/admission.go

```diff
@@ -0,0 +1,169 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package webhook checks a webhook for configured operation admission
+package webhook
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/util/webhook"
+
+	"k8s.io/kubernetes/pkg/api"
+	admissionv1alpha1 "k8s.io/kubernetes/pkg/apis/admission/v1alpha1"
+	kubeapiserveradmission "k8s.io/kubernetes/pkg/kubeapiserver/admission"
+
+	// install the clientgo admissiony API for use with api registry
+	_ "k8s.io/kubernetes/pkg/apis/admission/install"
+)
+
+var (
+	groupVersions = []schema.GroupVersion{
+		admissionv1alpha1.SchemeGroupVersion,
+	}
+)
+
+func init() {
+	kubeapiserveradmission.Plugins.Register("GenericAdmissionWebhook", func(configFile io.Reader) (admission.Interface, error) {
+
+		var gwhConfig struct {
+			WebhookConfig GenericAdmissionWebhookConfig `json:"webhook"`
+		}
+
+		d := yaml.NewYAMLOrJSONDecoder(configFile, 4096)
+		err := d.Decode(&gwhConfig)
+
+		if err != nil {
+			return nil, err
+		}
+
+		plugin, err := NewGenericAdmissionWebhook(&gwhConfig.WebhookConfig)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return plugin, nil
+	})
+}
+
+// NewGenericAdmissionWebhook returns a generic admission webhook plugin.
+func NewGenericAdmissionWebhook(config *GenericAdmissionWebhookConfig) (admission.Interface, error) {
+	err := normalizeConfig(config)
+
+	if err != nil {
+		return nil, err
+	}
+
+	gw, err := webhook.NewGenericWebhook(api.Registry, api.Codecs, config.KubeConfigFile, groupVersions, config.RetryBackoff)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &GenericAdmissionWebhook{
+		Handler: admission.NewHandler(admission.Connect, admission.Create, admission.Delete, admission.Update),
+		webhook: gw,
+		rules:   config.Rules,
+	}, nil
+}
+
+// GenericAdmissionWebhook is an implementation of admission.Interface.
+type GenericAdmissionWebhook struct {
+	*admission.Handler
+	webhook *webhook.GenericWebhook
+	rules   []Rule
+}
+
+// Admit makes an admission decision based on the request attributes.
+func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) (err error) {
+	// The default rule will skip processing of the admission request
+	matched := DefaultRule
+	matchedIndex := -1
+
+	// Process all declared rules to attempt to find a match
+	for i, rule := range a.rules {
+		if Matches(rule, attr) {
+			glog.V(2).Infof("rule at index %d matched request", i)
+			matched = rule
+			matchedIndex = i
+
+			break
+		}
+	}
+
+	if matchedIndex == -1 {
+		glog.V(2).Infof("no rules matched request, using default")
```

Comment:
This is misleading (or will be to people reading the logs and not the source code), since default means skip. Do you really need the concept of "default"?